lib/pos_blockchain.go

```diff
@@ -968,6 +968,50 @@ func (bc *Blockchain) getHighestCommittedBlock() (*BlockNode, int) {
 	return nil, -1
 }
 
+func (bc *Blockchain) GetSafeBlocks() ([]*BlockNode, error) {
+	// First get committed tip.
+	committedTip, idx := bc.getHighestCommittedBlock()
+	if idx == -1 || committedTip == nil {
+		return nil, errors.New("GetSafeBlocks: No committed blocks found")
+	}
+	// Now get all blocks from the committed tip to the best chain tip.
+	safeBlocks := []*BlockNode{committedTip}
+	maxHeightWithSafeBlocks := bc.getMaxSequentialBlockHeightAfter(uint64(committedTip.Height))
+	for ii := uint64(committedTip.Height + 1); ii < maxHeightWithSafeBlocks+1; ii++ {
+		blockNodes, exists := bc.blockIndexByHeight[ii]
+		if !exists {
+			continue
+		}
+		hasSeenValidatedBlockAtThisHeight := false
+		for _, blockNode := range blockNodes {
+			// TODO: Are there other conditions we should consider?
+			if blockNode.IsValidated() {
+				hasSeenValidatedBlockAtThisHeight = true
+				safeBlocks = append(safeBlocks, blockNode)
+			}
+		}
+		if !hasSeenValidatedBlockAtThisHeight {
+			break
+		}
+	}
+	return safeBlocks, nil
+}
+
+// getMaxSequentialBlockHeightAfter returns the max sequential block height after the starting height.
+// If the blockIndexByHeight does not have any blocks at a certain height, we know that any blocks
+// at a later height are not valid.
+func (bc *Blockchain) getMaxSequentialBlockHeightAfter(startingHeight uint64) uint64 {
+	// TODO: Are there further optimizations we can make here? What if we hit a
+	// height where there are blocks but they're all validate failed.
+	hasBlocksAtCurrentHeight := true
+	maxSequentialHeightWithBlocks := startingHeight
+	for currentHeight := startingHeight; hasBlocksAtCurrentHeight; currentHeight++ {
+		maxSequentialHeightWithBlocks = currentHeight
+		_, hasBlocksAtCurrentHeight = bc.blockIndexByHeight[currentHeight]
```

Comment:
This is fine. It'll be safer to have a wrapper function to check if we have blocks at a certain height like in this comment: https://github.com/deso-protocol/core/pull/785/files#r1383546779